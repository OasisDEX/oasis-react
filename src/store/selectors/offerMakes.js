import { createSelector, createStructuredSelector } from "reselect";
import { formValueSelector } from "redux-form/immutable";
import balances from "./balances";
import transactions from "./transactions";
import markets from "./markets";
import web3 from "../../bootstrap/web3";
import {
  MAKE_BUY_OFFER,
  MAKE_BUY_OFFER_FORM_NAME,
  MAKE_SELL_OFFER,
  MAKE_SELL_OFFER_FORM_NAME
} from "../../constants";
import reselect from "../../utils/reselect";
import {
  offerMakeToFormName,
  formNameToOfferMake
} from '../../utils/offers/offerMakeToFormName'
import tokens from "./tokens";
import network from "./network";
import { fromJS, Map } from "immutable";
import limits from "./limits";
import isNumeric from "../../utils/numbers/isNumeric";
import {memoize} from 'lodash';
import { offerMakeTypeToOfferListName } from '../../utils/offers/offerMakeTypeToOfferListName'
import { getOfferPrice } from '../../utils/offers/getOfferPrice'

const offerMakes = state => state.get("offerMakes");

const formFields = createStructuredSelector({
  [MAKE_BUY_OFFER_FORM_NAME]: reselect.formFieldsSelector(MAKE_BUY_OFFER_FORM_NAME, 'volume', 'price', 'total'),
  [MAKE_SELL_OFFER_FORM_NAME]: reselect.formFieldsSelector(MAKE_SELL_OFFER_FORM_NAME, 'volume', 'price', 'total'),
})

const currentFormValues = createSelector(
  formFields,
  formValues => memoize(formName => formValues[formName])
);

//TODO: move to utils
const toWeiString = amount =>
  web3
  .toBigNumber(web3.toWei(amount))
  .ceil()
  .toString();

const selector = createSelector(
  state => state.a,
  state => state.b,
  (a, b) => ({
    c: a * 2,
    d: b * 3
  })
);

const hasExceededGasLimit = createSelector(
  offerMakes,
  s => Boolean(s.get('exceededGasLimit'))
);


const activeTradingPairOffersData = createSelector(
  rootState => rootState.get("offers"),
  tokens.activeTradingPair,
  (state, activeTradingPair) => state.getIn(["offers", Map(activeTradingPair)])
);

const getNewOfferRankIndex = createSelector(
  (...args) => args[1],
  currentFormValues,
  activeTradingPairOffersData,
  (
    offerFormName,
    currentFormValues,
    currentTradingPairOffersData = fromJS({ buyOffers: [], sellOffers: [] })
  ) => {
    const { price } = currentFormValues(offerFormName);
    if (!parseFloat(String(price))) {
      return;
    }

    const priceBN = web3.toBigNumber(price);
    const formMakeType = formNameToOfferMake(offerFormName);
    const offersSortedByPriceAsc = currentTradingPairOffersData
    .get(offerMakeTypeToOfferListName(formMakeType))
    .map(offer => ({
      price: getOfferPrice(offer, formMakeType),
      offerId: offer.id
    }))
    .sort(
      ({ price: priceA }, { price: priceB }) => (priceA.gte(priceB) ? 1 : -1)
    )
    .map((d, idx) => ({ ...d, idx }));

    for (const { price, offerId, idx } of offersSortedByPriceAsc) {
      switch (priceBN.comparedTo(price)) {
        case -1:
          // console.log('offerId', offerId.toString());
          return offerId;
        case 0: {
          if (idx > 0) {
            const { offerId } = offersSortedByPriceAsc.get(idx - 1);
            // console.log('offerId', offerId.toString());
            return offerId;
          } else {
            const { offerId } = offersSortedByPriceAsc.first();
            // console.log('offerId', offerId.toString());
            return offerId;
          }
        }
      }
    }

    if (offersSortedByPriceAsc.count()) {
      // console.log('last', offersSortedByPriceAsc.last().offerId.toString());
      return offersSortedByPriceAsc.last().offerId;
    } else {
      // console.log('last', 0);
      return 0;
    }
  }
);

const activeOfferMakePure = createSelector(
  (...args) => args[1], //provides original selector argument
  tokens.activeTradingPair,
  network.tokenAddresses,
  currentFormValues,
  (rootState, offerMakeFormName) =>
    getNewOfferRankIndex(rootState, offerMakeFormName),
  (
    offerMakeFormName,
    activeTradingPair,
    tokenAddresses,
    currentFormValues,
    newOfferRankIndex
  ) => {
    const { total, volume } = currentFormValues(offerMakeFormName);
    const { baseToken, quoteToken } = activeTradingPair.toJS
      ? activeTradingPair.toJS()
      : activeTradingPair;
    const offerMakeType = formNameToOfferMake(offerMakeFormName);
    console.assert(
      offerMakeType,
      `Wrong offerMakeFormName: ${offerMakeFormName}`
    );
    const buyToken = offerMakeType === MAKE_BUY_OFFER ? baseToken : quoteToken;
    const sellToken = offerMakeType === MAKE_BUY_OFFER ? quoteToken : baseToken;

    return fromJS({
      baseToken,
      quoteToken,
      buyToken,
      sellToken,
      newOfferRankIndex,
      sellTokenAddress: tokenAddresses.get(sellToken),
      buyTokenAddress: tokenAddresses.get(buyToken),
      offerData: {
        payAmount: toWeiString(
          offerMakeType === MAKE_BUY_OFFER ? total : volume
        ),
        buyAmount: toWeiString(
          offerMakeType === MAKE_BUY_OFFER ? volume : total
        )
      }
    });
  }
);


// const activeOfferMake = activeOfferMakePure;

const activeOfferMakeType = createSelector(offerMakes, state =>
  state.get("activeOfferMakeType")
);

const activeOfferMakeOfferData = createSelector(activeOfferMakePure, state =>
  state.getIn("offerData")
);

const activeOfferMakeOfferOwner = createSelector(offerMakes, state =>
  state.getIn(["activeOfferMake", "offerData", "owner"])
);

const activeOfferMakeBuyToken = createSelector(activeOfferMakePure, state =>
  state.get("buyToken")
);

const activeOfferMakeSellToken = createSelector(activeOfferMakePure, state =>
  state.get("sellToken")
);

const isOfferMakeModalOpen = createSelector(
  offerMakes,
  reselect.getProps,
  (state) => memoize(offerMakeType =>
    state.getIn([
      offerMakeToFormName(offerMakeType),
      "isOfferMakeModalOpen"
    ])
  )
);

const makeFormValuesSelector = formName => {
  return formValueSelector(formName);
};

const hasSufficientTokenAmount = createSelector(
  balances.tokenBalances,
  currentFormValues,
  tokens.activeTradingPair,
  (
    tokenBalances,
    currentFormValues,
    { baseToken, quoteToken }
  ) => memoize(activeOfferMakeType => {
    const { total, volume } = currentFormValues(offerMakeToFormName(activeOfferMakeType));
    if (!volume) {
      return true;
    } else {
      switch (activeOfferMakeType) {
        case MAKE_BUY_OFFER: {
          const usersSellTokenBalanceBN = web3.toBigNumber(
            tokenBalances.get(quoteToken)
          );
          return usersSellTokenBalanceBN.gte(web3.toWei(total));
        }
        case MAKE_SELL_OFFER: {
          const usersSellTokenBalanceBN = web3.toBigNumber(
            tokenBalances.get(baseToken)
          );
          return usersSellTokenBalanceBN.gte(web3.toWei(volume));
        }
      }
    }
  })
);

const isVolumeOrPriceEmptyOrZero = createSelector(
  currentFormValues,
  (currentFormValues) => memoize(offerMakeType => {
    const { volume, price } = currentFormValues(offerMakeToFormName(offerMakeType));
    return !isNumeric(price) ||
      web3.toBigNumber(price).lte(0) ||
      (!isNumeric(volume) || web3.toBigNumber(volume).lte(0))
  })
);


const tokenMinSellLimitInWeiByOfferType = createSelector(
  tokens.activeTradingPair,
  limits.tokenMinSellLimitInWei,
  (activeTradingPair, tokenMinSellLimitInWei) => memoize(offerMakeType => {
    let tokenName = null;
    const activeTradingPairMap = Map(activeTradingPair);
    if (!activeTradingPairMap.has('baseToken') && activeTradingPairMap.has('quoteToken')) {
      return null
    }
    switch (offerMakeType) {
      case MAKE_BUY_OFFER:
        tokenName = activeTradingPairMap.get("quoteToken");
        break;
      case MAKE_SELL_OFFER:
        tokenName = activeTradingPairMap.get("baseToken");
        break;
      default:
        throw new Error("No offer make type provided!");
    }
    return tokenMinSellLimitInWei(tokenName);
  }));

const isOfferBelowLimit = createSelector(
  tokenMinSellLimitInWeiByOfferType,
  currentFormValues,
  (tokenMinSellLimitInWeiByOfferType, currentFormValues) => memoize(offerMakeType => {
    const { total, volume } = currentFormValues(offerMakeToFormName(offerMakeType));
    const tokenMinSellLimitInWei = tokenMinSellLimitInWeiByOfferType(offerMakeType);
    if (tokenMinSellLimitInWei === null) {
      return null;
    }
    let amount = null;
    switch (offerMakeType) {
      case MAKE_BUY_OFFER:
        amount = total;
        break;
      case MAKE_SELL_OFFER:
        amount = volume;
        break;
      default:
        throw new Error("No offer make type provided!");
    }

    if (
      isNumeric(tokenMinSellLimitInWei) &&
      isNumeric(amount) &&
      parseFloat(amount) > 0
    ) {
      return web3.toBigNumber(web3.fromWei(tokenMinSellLimitInWei)).gt(amount);
    }
  })
);

const isTotalOverTheTokenLimit = createSelector(
  limits.activeTradingPairQuoteTokenMaxLimitInWei,
  (rootState, offerMakeType) =>
    makeFormValuesSelector(offerMakeToFormName(offerMakeType))(
      rootState, "total",
    ),
  (tokenMaxSellLimitInWei, total) => {
    if (tokenMaxSellLimitInWei === null) {
      return null;
    } else if (
      isNumeric(tokenMaxSellLimitInWei) && isNumeric(total) && parseFloat(total) > 0
    ) {
      return web3.toBigNumber(web3.fromWei(tokenMaxSellLimitInWei)).lte(total);
    }
  }
);

const checkTokenEnabled = createSelector(
  s => s,
  (...args) => args[1],
  (...args) => args[2],
  (rootState, offerType, skipTokenEnabledCheck) => {
    return (
      skipTokenEnabledCheck ||
      balances.tokenAllowanceStatusForActiveMarket(rootState, {
        tokenName: activeOfferMakeSellToken(
          rootState,
          offerMakeToFormName(offerType)
        )
      })
    );
  },
);

const canMakeOffer = createSelector(
  hasSufficientTokenAmount,
  rootState => transactions.canSendTransaction(rootState),
  markets.isBuyEnabled,
  checkTokenEnabled,
  isVolumeOrPriceEmptyOrZero,
  isOfferBelowLimit,
  isTotalOverTheTokenLimit,
  (s, offerMakeType) => offerMakeType,
  (
    hasSufficientTokenAmount,
    canSendTransaction,
    isBuyEnabled,
    isTokenEnabled,
    isVolumeOrPriceEmptyOrZero,
    isBelowLimit,
    isOverTheLimit,
    offerMakeType,
  ) => {
    if (
      isVolumeOrPriceEmptyOrZero(offerMakeType) ||
      !hasSufficientTokenAmount(offerMakeType) ||
      // !canSendTransaction ||
      !isBuyEnabled ||
      !isTokenEnabled ||
      isBelowLimit(offerMakeType) ||
      isOverTheLimit
    ) {
      return false;
    } else {
      return hasSufficientTokenAmount(offerMakeType);
    }
  }
);

const isOfferActive = createSelector(offerMakes, state =>
  state.get("isOfferActive")
);

const transactionGasCostEstimate = createSelector(offerMakes, state =>
  state.get("transactionGasCostEstimate")
);

const checkingIfOfferIsActive = createSelector(offerMakes, state =>
  state.get("checkingIfOfferActive")
);

const gasEstimatePending = createSelector(offerMakes, state =>
  state.get("transactionGasCostEstimatePending")
);

const getActiveOfferMakeAllowanceStatus = createSelector(
  (rootState, offerType) => {
    return balances.tokenAllowanceStatusForActiveMarket(rootState, {
      tokenName: activeOfferMakeSellToken(
        rootState,
        offerMakeToFormName(offerType)
      )
    });
  },
  status => Boolean(status)
);

const isMakeBuyOfferPriceSet = createSelector(
  rootState => currentFormValues(rootState)(MAKE_BUY_OFFER_FORM_NAME),
  ({ price }) => Boolean(price)
);
const isMakeSellOfferPriceSet = createSelector(
  rootState => currentFormValues(rootState)(MAKE_SELL_OFFER_FORM_NAME),
  ({ price }) => Boolean(price)
);


export default {
  state: offerMakes,
  selector,
  activeOfferMakePure,
  activeOfferMakeType,
  activeOfferMakeOfferData,
  isOfferMakeModalOpen,
  activeOfferMakeBuyToken,
  activeOfferMakeSellToken,
  activeOfferMakeOfferOwner,
  canMakeOffer,
  hasSufficientTokenAmount,
  transactionGasCostEstimate,
  isVolumeOrPriceEmptyOrZero,
  isOfferActive,
  checkingIfOfferIsActive,
  currentFormValues,
  gasEstimatePending,
  getActiveOfferMakeAllowanceStatus,
  isMakeBuyOfferPriceSet,
  isMakeSellOfferPriceSet,
  isOfferBelowLimit,
  hasExceededGasLimit,
  isTotalOverTheTokenLimit
};
