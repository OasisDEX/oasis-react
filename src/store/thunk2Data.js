export default () =>  {
  let first = true;
  return store => next => action => {

    if(first) {
      first = false;
      return next(action);
    }

    if (action instanceof Array) {
      const [thunk, ...args] = action;
      return next({type: 'deferred-thunk', payload: {thunk, thunkName: thunk.name, args}});
    } else if (typeof action === 'function') {
      return next({type: 'thunk', payload: {thunk: action, thunkName: action.name}});
    } else {
      return next(action);
    }
  }
}