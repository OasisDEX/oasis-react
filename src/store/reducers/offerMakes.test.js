import thunk from 'redux-thunk';
import thunk2Data from '../thunk2Data';
import configureMockStore from 'redux-mock-store'
import {MAKE_BUY_OFFER, MAKE_SELL_OFFER} from '../../constants'
import { fromJS } from 'immutable';

import offerMakes from './offerMakes'

import each from 'jest-each';

describe('make offer form', () => {
    const testCases = [
        ["reasonable amount", "100", MAKE_BUY_OFFER],
        ["zero", "0", MAKE_BUY_OFFER],
        ["invalid", "a", MAKE_BUY_OFFER],
        ["negative number", "-1", MAKE_BUY_OFFER],
        ["reasonable amount", "100", MAKE_SELL_OFFER],
        ["zero", "0", MAKE_SELL_OFFER],
        ["invalid", "a", MAKE_SELL_OFFER],
        ["negative number", "-1", MAKE_SELL_OFFER]
    ]

    each(testCases).describe("volumeFieldValueChangedEpic", (desc, volume, offerMakeType) => {
        test(`volume set to: ${desc}(${volume}) for: ${offerMakeType}`, () => {

            const store = configureMockStore([thunk2Data(), thunk])({});

            store.dispatch(offerMakes.actions.volumeFieldValueChangedEpic(
              offerMakeType, volume, {localFormValueSelector: () => () => 3}));

            expect(store.getActions()).toMatchSnapshot();
        });
    });

    each(testCases).describe("totalFieldValueChangedEpic", (desc, total, offerMakeType) => {
        test(`total set to: ${desc}(${total}) for: ${offerMakeType}`, () => {

          const store = configureMockStore([thunk2Data(), thunk])({});

          store.dispatch(offerMakes.actions.totalFieldValueChangedEpic(
            offerMakeType, total, {localFormValueSelector: () => () => 3}))

          expect(store.getActions()).toMatchSnapshot();
        });
    });


    each(testCases).describe("priceFieldChangedEpic", (desc, price, offerMakeType) => {
        test(`price set to: ${desc}(${price}) for: ${offerMakeType}`, () => {
          const store = configureMockStore([thunk2Data(), thunk])({});

          store.dispatch(offerMakes.actions.totalFieldValueChangedEpic(
            offerMakeType, price, {localFormValueSelector: () => () => 3}))

          expect(store.getActions()).toMatchSnapshot();
        });
    });
});

describe("make offer modal", () => {
    test("setOfferMakeModalOpenEpic", () => {
        const store = configureMockStore([thunk2Data(), thunk])({});
        store.dispatch(offerMakes.actions.setOfferMakeModalOpenEpic(MAKE_BUY_OFFER));
        expect(store.getActions()).toMatchSnapshot();
    });
    test("setOfferMakeModalClosedEpic", () => {
        const store = configureMockStore([thunk2Data(), thunk])({});
        store.dispatch(offerMakes.actions.setOfferMakeModalClosedEpic(MAKE_BUY_OFFER));
        expect(store.getActions()).toMatchSnapshot();
    })
});

each([[MAKE_BUY_OFFER], [MAKE_SELL_OFFER]]).test("buyMaxEpic", (offerMakeType) => {
    const store = configureMockStore([thunk2Data(), thunk])({});

    store.dispatch(offerMakes.actions.buyMaxEpic(offerMakeType, { activeQuoteTokenBalance: () => 33}));

    expect(store.getActions()).toMatchSnapshot();
});

each([[MAKE_BUY_OFFER], [MAKE_SELL_OFFER]]).test("sellMaxEpic", (offerMakeType) => {

  const store = configureMockStore([thunk2Data(), thunk])({});

  store.dispatch(offerMakes.actions.sellMaxEpic(
    offerMakeType,
    { activeBaseTokenBalance: () => 100,
      currentFormValues: () => () => ({price: 0.8})
    }));

  expect(store.getActions()).toMatchSnapshot();
});

each([[true], [false]]).test("updateTransactionGasCostEstimateEpic", (canMakeOffer) => {
  const store = configureMockStore([thunk2Data(), thunk])(
    fromJS({ transactions: {}})
  );

  store.dispatch(offerMakes.actions.updateTransactionGasCostEstimateEpic(
    MAKE_BUY_OFFER,
    { canMakeOffer: () => canMakeOffer,
      activeOfferMakePure: () => fromJS({
        sellTokenAddress: "0xd0a1e359811322d97991e03f863a0c30c2cf029c",
        buyTokenAddress: "0xaaf64bfcc32d0f15873a02163e7e500671a4ffcd",
        offerData: {
          payAmount: "1000000000000000000",
          buyAmount: "1000000000000000000",
        }
      })
    }));
  expect(store.getActions()).toMatchSnapshot();
});
