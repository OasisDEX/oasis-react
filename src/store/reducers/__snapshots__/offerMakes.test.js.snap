// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buyMaxEpic 1`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeBuyOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0.000000000000000033",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
        "0.000000000000000033",
      ],
      "thunk": [Function],
      "thunkName": "totalFieldValueChangedEpic",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`buyMaxEpic 2`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeSellOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0.000000000000000033",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
        "0.000000000000000033",
      ],
      "thunk": [Function],
      "thunkName": "totalFieldValueChangedEpic",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form priceFieldChangedEpic price set to: invalid(a) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form priceFieldChangedEpic price set to: invalid(a) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form priceFieldChangedEpic price set to: negative number(-1) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form priceFieldChangedEpic price set to: negative number(-1) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form priceFieldChangedEpic price set to: reasonable amount(100) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form priceFieldChangedEpic price set to: reasonable amount(100) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form priceFieldChangedEpic price set to: zero(0) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form priceFieldChangedEpic price set to: zero(0) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form totalFieldValueChangedEpic total set to: invalid(a) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form totalFieldValueChangedEpic total set to: invalid(a) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form totalFieldValueChangedEpic total set to: negative number(-1) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form totalFieldValueChangedEpic total set to: negative number(-1) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form totalFieldValueChangedEpic total set to: reasonable amount(100) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form totalFieldValueChangedEpic total set to: reasonable amount(100) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form totalFieldValueChangedEpic total set to: zero(0) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form totalFieldValueChangedEpic total set to: zero(0) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form volumeFieldValueChangedEpic volume set to: invalid(a) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeBuyOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form volumeFieldValueChangedEpic volume set to: invalid(a) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeSellOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form volumeFieldValueChangedEpic volume set to: negative number(-1) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeBuyOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form volumeFieldValueChangedEpic volume set to: negative number(-1) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeSellOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form volumeFieldValueChangedEpic volume set to: reasonable amount(100) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeBuyOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form volumeFieldValueChangedEpic volume set to: reasonable amount(100) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeSellOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form volumeFieldValueChangedEpic volume set to: zero(0) for: OFFER_MAKES/MAKE_BUY_OFFER 1`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeBuyOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer form volumeFieldValueChangedEpic volume set to: zero(0) for: OFFER_MAKES/MAKE_SELL_OFFER 1`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeSellOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
      ],
      "thunk": [Function],
      "thunkName": "debounced",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`make offer modal setOfferMakeModalClosedEpic 1`] = `
Array [
  Object {
    "payload": Object {
      "formName": "makeBuyOffer",
    },
    "type": "OFFER_MAKES/SET_MODAL_CLOSED",
  },
  Object {
    "type": "OFFER_MAKES/RESET_ACTIVE_OFFER_MAKE_TYPE",
  },
  Object {
    "type": "OFFER_MAKES/RESET_TRANSACTION_GAS_ESTIMATE",
  },
]
`;

exports[`make offer modal setOfferMakeModalOpenEpic 1`] = `
Array [
  Object {
    "payload": "OFFER_MAKES/MAKE_BUY_OFFER",
    "type": "OFFER_MAKES/SET_ACTIVE_OFFER_MAKE_TYPE",
  },
  Object {
    "payload": Object {
      "thunk": [Function],
      "thunkName": "",
    },
    "type": "thunk",
  },
  Object {
    "payload": Object {
      "formName": "makeBuyOffer",
    },
    "type": "OFFER_MAKES/SET_MODAL_OPEN",
  },
]
`;

exports[`sellMaxEpic 1`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeBuyOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0.000000000000000033",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_BUY_OFFER",
        "0.000000000000000033",
      ],
      "thunk": [Function],
      "thunkName": "totalFieldValueChangedEpic",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`sellMaxEpic 2`] = `
Array [
  Object {
    "meta": Object {
      "field": "total",
      "form": "makeSellOffer",
      "persistentSubmitErrors": undefined,
      "touch": undefined,
    },
    "payload": "0.000000000000000033",
    "type": "@@redux-form/CHANGE",
  },
  Object {
    "payload": Object {
      "args": Array [
        "OFFER_MAKES/MAKE_SELL_OFFER",
        "0.000000000000000033",
      ],
      "thunk": [Function],
      "thunkName": "totalFieldValueChangedEpic",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`updateTransactionGasCostEstimateEpic 1`] = `
Array [
  Object {
    "payload": Object {
      "args": Array [
        "1000000000000000000",
        "0xd0a1e359811322d97991e03f863a0c30c2cf029c",
        "1000000000000000000",
        "0xaaf64bfcc32d0f15873a02163e7e500671a4ffcd",
        undefined,
      ],
      "thunk": [Function],
      "thunkName": "actionCreator",
    },
    "type": "deferred-thunk",
  },
]
`;

exports[`updateTransactionGasCostEstimateEpic 2`] = `
Array [
  Object {
    "type": "OFFER_MAKES/RESET_TRANSACTION_GAS_ESTIMATE",
  },
]
`;
